classes/Deck.js

```diff
@@ -1,22 +1,115 @@
-import { PlayingCard } from './PlayingCard'
+import { PlayingCard, SUITES, RANKS } from './playing-card'
 
 class Deck {
-  constructor(suites, ranks) {
+  constructor(cards) {
+    this.cards = cards
+  }
+
+  static from(suites, ranks) {
     const cards = []
+    let i = 0
     for (const suite of suites) {
       for (const rank of ranks) {
-        cards.push(new PlayingCard(suite, rank))
+        cards.push(new PlayingCard(suite, rank, i++))
       }
     }
-    this.cards = cards
+    return cards
+  }
+
+  print() {
+    for (const card of this.cards) {
+      console.log(card.toString())
+    }
+  }
+
+  drawTopCard() {
+    this.validateNotEmpty('drawTopCard')
+
+    const card = this.cards.pop()
+    this.checkIsEmpty()
+
+    return card
+  }
+  
+  drawBottomCard() {
+    this.validateNotEmpty('drawBottomCard')
+
+    const card = this.cards.shift()
+    this.checkIsEmpty()
+
+    return card
+  }
+
+  insertCardTop(card) {
+    this.cards.push(card)
+  }
+
+  insertCardBottom(card) {
+    this.cards.unshift(card)
+  }
+
+  checkIsEmpty() {
+    this.isEmpty = this.cards.length === 0
+  }
+
+  validateNotEmpty(action) {
+    if (this.isEmpty) throw new Error(`Illegal action "${action}": Deck is empty`)
+  }
+
+  cut() {
+    const [ top, bottom ] = this.split()
+
+    this.cards = bottom.cards.concat(top.cards)
+  }
```

Comment:
I would move this below `split()` just because I was very confused about when you used `this.split()` when I hadn't seen it yet and I didn't associate it correctly with a method of this class. Like it was dumb of me obviously, but it probably would be less likely to happen had it been defined afterwards